src/query/src/dist_plan/merge_scan.rs

```diff
@@ -150,40 +150,36 @@ impl MergeScanExec {
         let metric = MergeScanMetric::new(&self.metric);
 
         let stream = try_stream! {
+            let _finish_timer = metric.finish_time().timer();
             for peer in peers {
                 let client = clients.get_client(&peer).await;
                 let database = Database::new(&table.catalog_name, &table.schema_name, client);
-                let _timer = metric.grpc_time().timer();
+
+                let mut ready_timer = metric.ready_time().timer();
+                let mut first_consume_timer = metric.first_consume_time().timer();
+                let mut stopped = false;
+
                 let output: Output = database
                     .logical_plan(substrait_plan.clone(), trace_id)
                     .await
                     .context(RemoteRequestSnafu)
                     .map_err(BoxedError::new)
                     .context(ExternalSnafu)?;
 
-                match output {
-                    Output::AffectedRows(_) => {
-                        Err(BoxedError::new(
-                            UnexpectedOutputKindSnafu {
-                                expected: "RecordBatches or Stream",
-                                got: "AffectedRows",
-                            }
-                            .build(),
-                        ))
-                        .context(ExternalSnafu)?;
-                    }
-                    Output::RecordBatches(record_batches) => {
-                        for batch in record_batches.into_iter() {
-                            metric.record_output_batch_rows(batch.num_rows());
-                            yield Self::remove_metadata_from_record_batch(batch);
-                        }
-                    }
-                    Output::Stream(mut stream) => {
-                        while let Some(batch) = stream.next().await {
-                            let batch = batch?;
-                            metric.record_output_batch_rows(batch.num_rows());
-                            yield Self::remove_metadata_from_record_batch(batch);
-                        }
+                let Output::Stream(mut stream) = output else {
+                    unreachable!()
+                };
+
+                ready_timer.stop();
+
+                while let Some(batch) = stream.next().await {
+                    let batch = batch?;
+                    metric.record_output_batch_rows(batch.num_rows());
+                    yield Self::remove_metadata_from_record_batch(batch);
+
+                    if !stopped {
```

Comment:
I think we can get rid of `stopped` like this: make `first_consume_timer` an `Option`, then `if let Some(t) = first_consume_timer.take() { t.stop() }`.